configure.ac

```diff
@@ -443,21 +443,34 @@ if test "x$with_dl" != "xno"; then
   AC_CHECK_LIB(dl, dlopen)
 fi
 
+AC_DEFUN([LIBATOMIC_CHECKER],[
+  AC_LINK_IFELSE([
+    AC_LANG_SOURCE([[
+      #include <atomic>
+      struct LockRequired { char data[256]; };
+      int
+      main() {
+          return std::atomic<LockRequired>{}.is_lock_free();
+      }
+  ]])],[
+    AC_MSG_RESULT(yes)
+    libatomic_checker_result="yes"],[
+    AC_MSG_RESULT(no)
+    libatomic_checker_result="no"
+])])
+
 ## check for atomics library before anything that might need it
 SQUID_STATE_SAVE(LIBATOMIC)
-AC_MSG_CHECKING(library containing __atomic_is_lock_free)
-LIBS="-latomic"
-AC_LINK_IFELSE(
-  [AC_LANG_SOURCE([[
-    int main(void) {
-        char lockRequired[256];
-        return __atomic_is_lock_free(sizeof(lockRequired),&lockRequired);
-    }
-  ]])],
-  [AC_MSG_RESULT(-latomic)
-   ATOMICLIB='-latomic'],
-  [AC_MSG_RESULT(no)]
-)
+AC_MSG_CHECKING(whether linking without -latomic works)
+LIBATOMIC_CHECKER()
+if test "x$libatomic_checker_result" = "xno"; then
+  AC_MSG_CHECKING(whether linking with -latomic works)
+  LIBS="$LIBS -latomic"
+  LIBATOMIC_CHECKER()
+  if test "x$libatomic_checker_result" = "xyes"; then
+    ATOMICLIB="-latomic"
+  fi
```

Comment:
AFAICT, the (currently absent) `else` clause should either terminate ./configure with an error (Amos suggestion IIRC) or at least restore polluted `LIBS` (to emulate `AC_SEARCH_LIBS`' never-failing behavior).

Reply:
>  at least restore polluted LIBS

This should be perfomed by SQUID_STATE_SAVE/SQUID_STATE_ROLLBACK pair.

Reply:
Sorry I missed them in the diff.